### ISSUE TITLE
ISTIO related deployments are deployed to wrong namespace

### ISSUE THREAD

**Issue by gabrielwen** [kubeflow/manifests#99]
Should be in `istio-system`, but it's deployed to `kubeflow` instead.

example:
```
Name:             iap-ingress-envoy-ingress
Namespace:        kubeflow
Address:
Default backend:  default-http-backend:80 (10.16.0.10:8080)
Rules:
  Host                                                        Path  Backends
  ----                                                        ----  --------
  kustomize-testing.endpoints.gabrielwen-learning.cloud.goog
                                                              /*   istio-ingressgateway:80 (<none>)
Annotations:
  certmanager.k8s.io/issuer:                    letsencrypt-prod
  ingress.kubernetes.io/ssl-redirect:           true
  kubernetes.io/ingress.global-static-ip-name:  kustomize-testing-ip
  kubernetes.io/tls-acme:                       true
Events:
  Type     Reason     Age                 From                     Message
  ----     ------     ----                ----                     -------
  Normal   ADD        16m                 loadbalancer-controller  kubeflow/iap-ingress-envoy-ingress
  Warning  Translate  59s (x20 over 16m)  loadbalancer-controller  error while evaluating the ingress spec: could not find service "kubeflow/istio-ingressgateway"
```

**Comment by kkasravi** [comment]
@gabrielwen 

Here is my suggestion, which is similar to how I'm implementing 'applications'.
I thought I would outline the approach here, the implementation is in progress (can be aborted/changed/discussed).

When useIstio == true (cli option)

- kubeflow bootstrap/config changes
   - add a config/overlays/istio
   - this overlay adds new components istio-install, istio to app.yaml 
   - this overlay adds enables istio overlays in components with virtual-services
   - this overlay sets the namespace of iap-ingress, istio-install, istio to istio-system
   - remove useIstio, "false" in gcp components - useIstio, "true" is now overlay, "istio" in these components
     
- manifests changes
  - move dependencies/istio/* to  gcp/{istio-install, istio}
  - create overlays/istio for components with virtual-services, move their virtual-services.yaml there

pros:
- moves most of istio into the generate phase so it can be examined prior to apply
- moves the definition of 'istio-system' into app.yaml instead of in the code (IstioNamespace)
- cleaner manifests since we're isolating where the istio-system namespace is defined and it only needs to be in kustomization.yaml
- isolates the virtual-service into an overlay for affected components
- moves most of the istio specific code in gcp.go to manifests
- kustomize does istio based on params in the generated app.yaml - versus having gcp.go do it.

cons:
- fair amount of reshuffling 
- possible order-of-execution issues

**Comment by gabrielwen** [comment]
/cc @lluunn @jlewi 

A few things:
- We will remove ambassador in v0.6 so I don't think we need to consider `useIstio == false`.
- I think there is pushback on moving dependencies/istio to manifests but I'll leave this for @lluunn and @jlewi to decide.  My only concern is that although it'd be good that we can keep all manifests in the same place, I don't see an obvious benefit to do this.  Plus, the biggest concern to me is kustomize will rewrite ISTIO manifests which in any case shouldn't.
- I think `namespace: istio-system` is set by ISTIO team and they have this assumption in many other components.  Which means, we shouldn't change this in most of the cases.
- I kinda understand how overlay can help on this issue, but could you give me an example on how file layout would like?  Is it something like this:
```
gcp/base/...
<components-wo-istio>/base/...
overlay/base/<component-w-istio>-virtual-service.yaml
      ...            /<component-w-istio>.yaml
```

**Comment by kkasravi** [comment]
thanks for weighing in. I'm about to submit the PR(s) - also answering comments below. We can probably continue within the scope of the PR(s).

> We will remove ambassador in v0.6 so I don't think we need to consider useIstio == false.

+1

> I think there is pushback on moving dependencies/istio to manifests but I'll leave this for @lluunn and @jlewi to decide. My only concern is that although it'd be good that we can keep all manifests in the same place, I don't see an obvious benefit to do this. Plus, the biggest concern to me is kustomize will rewrite ISTIO manifests which in any case shouldn't.

one goal is to minimize special processing in kfctl that could be done by kustomize Another argument is the yaml is rendered at generate time. The only kustomize processing would be adding namespace - it's possible to do more (labels, nameprefix) but we could/should just do this initially. 

> I think namespace: istio-system is set by ISTIO team and they have this assumption in many other components. Which means, we shouldn't change this in most of the cases.

yes - i'm in agreement. Though we wanted it to be easy to change the namespace if useIstio==false?

> I kinda understand how overlay can help on this issue, but could you give me an example on how file layout would like? 

sure 
```
.
└── manifests
    ├── argo
    │   ├── base
    │   └── overlays
    │       └── istio
    ├── common
    │   └── centraldashboard
    │       ├── base
    │       └── overlays
    │           └── istio
    ├── gcp
    │   ├── iap-ingress
    │   │   └── base
    │   ├── istio
    │   │   └── base
    │   └── istio-install
    │       └── base
    ├── jupyter
    │   └── jupyter-web-app
    │       ├── base
    │       └── overlays
    │           └── istio
    ├── kubebench
    │   ├── base
    │   └── overlays
    │       └── istio
    ├── pipeline
    │   ├── pipelines-ui
    │   │   ├── base
    │   │   └── overlays
    │   │       └── istio
    │   └── pipelines-viewer
    │       ├── base
    │       └── overlays
    │           └── istio
    ├── tensorboard
    │   ├── base
    │   └── overlays
    │       └── istio
    └── tf-training
        └── tf-job-operator
            ├── base
            └── overlays
                └── istio
```

**Comment by jlewi** [comment]
@kkasravi Is this fixed?  Do we have an E2E verifying that kustomize + ISTIO is working?

**Comment by kkasravi** [comment]
yes we are running kfctl e2e tests on master when we merged #102 and kubeflow/kubeflow#3375
